source/common/http/http2/metadata_decoder.cc

```diff
@@ -0,0 +1,110 @@
+#include "common/http/http2/metadata_decoder.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http2 {
+
+MetadataDecoder::MetadataDecoder(uint64_t stream_id) : stream_id_(stream_id) {
+  int rv = nghttp2_hd_inflate_new(&inflater_);
+  ASSERT(rv == 0);
+}
+
+MetadataDecoder::~MetadataDecoder() { nghttp2_hd_inflate_del(inflater_); }
+
+void MetadataDecoder::receiveMetadata(const uint8_t* data, size_t len) {
+  if (data == nullptr || len == 0) {
+    ENVOY_LOG(error, "No payload for the decoder to receive.");
+    return;
+  }
+
+  payload_.add(data, len);
```

Comment:
What's the plan on enforcing length limits on the read side?  I do think we want to either drop metadata or kill the connection if it gets too large.  Maybe TODO to handle eventually?

Reply:
That's a very good point! I added size limit in the decoder and updated the test.